modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuilder.java

```diff
@@ -0,0 +1,273 @@
+                                return completedFuture(null);
+                            }
+
+                            try {
+                                // At this point, we have a stable topology, each node of which has already applied all local updates.
+                                if (!localNodeConsistentId().equals(leaderWithTerm.leader().consistentId())) {
+                                    // TODO: IGNITE-19053 Must handle the change of leader
+                                    // TODO: IGNITE-19053 Add a test to change the leader even at the start of the task
+                                    return completedFuture(null);
+                                }
+
+                                List<RowId> batchRowIds = collectRowIdBatch();
+
+                                RowId nextRowId = getNextRowIdForNextBatch(batchRowIds);
+
+                                boolean finish = batchRowIds.size() < BUILD_INDEX_ROW_ID_BATCH_SIZE || nextRowId == null;
+
+                                // TODO: IGNITE-19053 Must handle the change of leader
+                                return raftGroupService.run(createBuildIndexCommand(batchRowIds, finish))
+                                        .thenRun(() -> {
+                                            if (!finish) {
+                                                assert nextRowId != null : createCommonTableIndexInfo();
+
+                                                buildIndexExecutor.submit(
+                                                        new BuildIndexTask(table, tableIndexView, partitionId, nextRowId)
+                                                );
+                                            }
+                                        });
+                            } finally {
+                                busyLock.leaveBusy();
+                            }
+                        }, buildIndexExecutor)
+                        .whenComplete((unused, throwable) -> {
+                            if (throwable != null) {
+                                LOG.error("Index build error: [{}]", throwable, createCommonTableIndexInfo());
+                            }
+                        });
+            } catch (Throwable t) {
+                LOG.error("Index build error: [{}]", t, createCommonTableIndexInfo());
+            } finally {
+                busyLock.leaveBusy();
+            }
+        }
+
+        private boolean isLocalNodeLeader(RaftGroupService raftGroupService) {
+            Peer leader = raftGroupService.leader();
+
+            assert leader != null : "tableId=" + table.tableId() + ", partitionId=" + partitionId;
+
+            return localNodeConsistentId().equals(leader.consistentId());
+        }
+
+        private List<RowId> createBatchRowIds(RowId lastBuiltRowId, int batchSize) {
+            MvPartitionStorage mvPartition = table.internalTable().storage().getMvPartition(partitionId);
```

Comment:
Why can't we store partition instance in the task, instead of retrieving it every time?

Reply:
Because we can easily get it from the `TableImpl`, there is no need to store it.

Reply:
Ok, as you wish